core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/progress/listener/ProgressListenerInvoker.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.progress.listener;
+
+import static software.amazon.awssdk.utils.FunctionalUtils.runAndLogError;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.core.progress.listener.ProgressListener;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * ProgressListenerInvoker class exposes thread safe methods that facilitate invocation of all ProgressListener methods
+ * corresponding to different stages of the request lifecycle
+ */
+@SdkInternalApi
+@ThreadSafe
+public class ProgressListenerInvoker implements ProgressListener {
+
+    private static final Logger log = Logger.loggerFor(ProgressListenerInvoker.class);
+
+    private final List<ProgressListener> listeners;
+    private final AtomicBoolean prepared = new AtomicBoolean();
+    private final AtomicBoolean headerSent = new AtomicBoolean();
+    private final AtomicBoolean headerReceived = new AtomicBoolean();
+    private final AtomicBoolean complete = new AtomicBoolean();
+
+    public ProgressListenerInvoker(List<ProgressListener> listeners) {
+        this.listeners = Validate.paramNotNull(listeners, "listeners");
+    }
+
+    @Override
+    public void requestPrepared(Context.RequestPrepared context) {
+        if (!prepared.getAndSet(true)) {
+            forEach(listener -> listener.requestPrepared(context));
+        }
+    }
+
+    @Override
+    public void requestHeaderSent(Context.RequestHeaderSent context) {
+        if (!headerSent.getAndSet(true)) {
+            forEach(listener -> listener.requestHeaderSent(context));
+        }
+    }
+
+    @Override
+    public void requestBytesSent(Context.RequestBytesSent context) {
+        forEach(listener -> listener.requestBytesSent(context));
```

Comment:
We do ThreadSafety checks for few api and not for others

Since "public interface ProgressListener" is public API should and If user is implementing the listeners is it a "MUST" that they need to handle the ThreadSafety  ? something we can discuss offline , just curious

Reply:
Operations like RequestHeaderSent, RequestPrepared are expected to be called exactly once for every request. Thread safety plus idempotence, we dont want repeated calling of the same API. But in the case of BytesSent, they are expected to be called multiple times per operation (with different snapshots). As discussed on Friday, Thread safety would not be an issue, technically, as this is stateless

Reply:
I am unsure if, this is technically a thread safe class. I guess since a few of the APIs are, we could call it so ?